spatialos-sdk/src/worker/schema/bundle.rs

```diff
@@ -0,0 +1,620 @@
+        let err = Schema_Json_GetLastError();
+        if err.is_null() {
+            "Unknown error in loading object".to_string()
+        } else {
+            cstr_to_string(err)
+        }
+    }
+
+    unsafe fn get_last_warning() -> Option<String> {
+        let maybe_warning = Schema_Json_GetLastWarning();
+        if maybe_warning.is_null() {
+            None
+        } else {
+            Some(cstr_to_string(maybe_warning))
+        }
+    }
+
+    pub fn convert_data_to_update(
+        &self,
+        component_id: ComponentId,
+        data: Owned<SchemaComponentData>,
+    ) -> std::result::Result<Owned<SchemaComponentUpdate>, String> {
+        let mut error: Option<String> = None;
+        unsafe {
+            let update_ptr = Schema_ConvertComponentDataIntoUpdate(
+                self.ptr.as_ptr(),
+                component_id,
+                data.into_raw(),
+                &mut error as *mut _ as *mut _,
+                Some(Bundle::record_error),
+            );
+
+            if update_ptr.is_null() {
+                return Err(error.unwrap_or_else(|| {
+                    "Unknown error occurred when converting data into update.".to_string()
+                }));
+            }
+
+            let update = SchemaComponentUpdate::from_raw_mut(update_ptr).to_owned();
+            Schema_DestroyComponentUpdate(update_ptr);
+
+            Ok(update)
+        }
+    }
+
+    extern "C" fn record_error(
+        user_data: *mut ::std::os::raw::c_void,
+        error: *const ::std::os::raw::c_char,
+    ) {
+        unsafe {
+            if error.is_null() {
+                return;
+            }
+
+            let data: &mut Option<String> = &mut *(user_data as *mut Option<String>);
+            let error = cstr_to_string(error);
+            *data = Some(error);
+        }
+    }
+}
+
+// SAFETY: It should be safe to send a `Bundle` between threads, so long as it's only ever accessed
+// from one thread at a time. It has unsychronized internal mutability (only storing the 'last' error
+// and warning) so it cannot be Sync.
+unsafe impl Send for Bundle {}
+
+impl Drop for Bundle {
+    fn drop(&mut self) {
+        unsafe {
+            Schema_Bundle_Destroy(self.ptr.as_ptr());
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::worker::{component::ComponentId, schema::*};
+    use serde_json::Value;
+    use std::{fs::File, io::Read, path::PathBuf};
+
+    const POSITION_COMPONENT_ID: ComponentId = 54;
+    const EPSILON: f64 = std::f64::EPSILON;
+
+    macro_rules! assert_delta {
+        ($x:expr, $y:expr, $d:expr) => {
+            assert!(($x - $y).abs() < $d)
+        };
+    }
+
+    #[test]
+    pub fn loading_bundle_succeeds_when_valid_bundle() {
+        assert!(read_bundle(true).is_ok(), "Valid bundle failed to load.")
+    }
+
+    // TODO: This test currently fails (but it shouldn't!)
+    //#[test]
+    pub fn loading_bundle_fails_when_invalid_bundle() {
```

Comment:
Did you want to get this test working as part of this PR? Or will that need to wait for a future update?

Reply:
A future update, this test revealed a bug in the underlying implementation. The fix was released in `14.5.0` so I'll update this TODO.